Feat/#55 user domain api 수정사항

DuckOriDuck opened the pull request:
## 🚀 개발 사항
리뷰해주신 수정사항 반영했고 작은 에러 몇 가지 수정했습니다
### 이슈 번호
- close 55

## 특이 사항 🫶 
없습니다.

DuckOriDuck committed:
[FEAT] User 도메인 엔티티, 리포지토리, enum 객체 추가

DuckOriDuck committed:
[FEAT] user 정보 가져오는 메소드 추가

DuckOriDuck committed:
[FEAT] controller, service용 리스폰스 엔티티 구현

DuckOriDuck committed:
[FEAT] service id기반으로 유저의 정보를 찾아 mypage용 리스폰스 엔티티 리턴해주는 메소드 생성

DuckOriDuck committed:
[FEAT]Usercontroller 생성, mypage 신호 요청 API 구현

DuckOriDuck committed:
FIX: MyPageAPIResponse 클래스가 직렬화 가능하도록 수정

DuckOriDuck committed:
FIX: 유저 인증정보 관련 임시 주석 처리

DuckOriDuck committed:
FIX: 빌더 어노테이션 위치 수정

DuckOriDuck committed:
[FIX] user 생성자 수정, delete 날 설정하는 메소드 구현

DuckOriDuck committed:
[FEAT] user delete 기능을 위한 response DTO 생성

DuckOriDuck committed:
[FEAT] 회원탈퇴 로직 메소드 deleteUserAcount 구현

DuckOriDuck committed:
[FEAT] 회원탈퇴 API 구현

DuckOriDuck committed:
[MIGRATION] Users테이블의 pk id에 auto increment 설정

DuckOriDuck committed:
[FEAT] service의 signup을 위한 request, response dto 생성

DuckOriDuck committed:
[FEAT] 회원가입 서비스로직 구현.
에러처리 로직 필요

DuckOriDuck committed:
[FEAT] 이메일이 존재하는지 확인하는 메소드 추가

DuckOriDuck committed:
[FEAT] repository에서 save() 메소드 쓸 시 created_at 설정하는 함수 설정

DuckOriDuck committed:
[FEAT] controller API signup을 위한 request, response dto 생성

DuckOriDuck committed:
[FEAT] 회원가입 기능 /signup path 추가

DuckOriDuck committed:
[FEAT] user관련 에러 추가
"E03-USER-001", "요청된 ID를 가진 유저가 존재하지 않습니다"
"E01-USER-001", "이미 존재하는 계정입니다."
"E01-USER-002", "탈퇴한 유저입니다"

DuckOriDuck committed:
[FEAT] 탈퇴한 유저인지 체크하기 위한 utility 클래스 UserValidator 추가

DuckOriDuck committed:
[FEAT] User 엔티티에 탈퇴 취소 로직 추가

DuckOriDuck committed:
[FEAT] User Exception 처리용 클래스 추가

DuckOriDuck committed:
[FEAT] UUsreException 오타 수정

DuckOriDuck committed:
[FEAT] Email로 유저 찾도록 함수 추가

DuckOriDuck committed:
[FEAT] 이미 존재하는 회원, 존재하지 않는 회원, 탈퇴한 회원에 대한 Exception들 추가

DuckOriDuck committed:
[FIX] 에러코드 이름 변경, USER_ACCOUNT_DISABLED 에러 401로 이동

DuckOriDuck committed:
[RENAME] 마이그레이션 파일 버젼 변경 V3->V7

DuckOriDuck committed:
[REFACTOR] DTO 클래스들에 final 추가

DuckOriDuck committed:
[FIX] 전부 다 이미 존재하는 계정으로 감지하는 에러 수정

DuckOriDuck committed:
[FEAT] 탈퇴 취소 로직 추가

DuckOriDuck committed:
[FIX] 유저 코드 수정
- 생성자 수정
- 상속 적용되도록 변수 수정
- 탈퇴/탈퇴취소 로직 수정

DuckOriDuck committed:
[REFACTOR] unSetDelete->restore, resotoreAccount()로 변경

DuckOriDuck committed:
[REFACTOR] 정적 팩토리 메소드로 변경

DuckOriDuck committed:
[REFACTOR] Validator 클래스 삭제, error throw는 유저단에서 하도록 설정

DuckOriDuck committed:
[FIX] value change를 Super class에서 해버리면 JPA가 값의 변경을 인지하지 못해서 DB에 write가 되지 않음.
write가 되게 하려면 repository를 사용해서 save하거나 entity 단에서 바꾸던가 해야하는데 entity에서 repository.save() 하는 것은 좋지 않은 방법 같아 BaseTimeEntity의 필드 변수들의 접근 제한을 protected로 설정.
- delete()는 혹시 모르니 놔두는 용

DuckOriDuck committed:
[REFACTOR] DeleteMapping path 수정

DuckOriDuck committed:
[FIX] repository 명시

DuckOriDuck committed:
[FIX] @Transactional 붙이니 해결됨.
setDelete LocadateTime 리턴하도록 설정

DuckOriDuck committed:
[FIX] 탈퇴시간 리턴 안하도록 수정

JIN-076 commented on src/main/java/io/dev/jobprep/common/base/BaseTimeEntity.java:
```diff
@@ -15,12 +15,12 @@ public class BaseTimeEntity {
+    protected LocalDateTime deletedAt;
 
-    protected void delete() {
+    protected void setDelete() {
```
메서드명에 setter를 사용하는 것은 권장하지 않습니다. 직관적으로 의미가 드러날 수 있는 메서드명을 사용하기 위해 `delete()` 로 명명했습니다. 다음 문서 읽어보시면 좋을 것 같습니다.

+ 추가적으로 이런 네이밍은 도메인 주도 개발 컨셉인 경우에 좀 더 핏하다고 볼 수 있을 것 같습니다. 

[setter 쓰지 말라고만 하고 가버리면 어떡해요](https://velog.io/@backfox/setter-%EC%93%B0%EC%A7%80-%EB%A7%90%EB%9D%BC%EA%B3%A0%EB%A7%8C-%ED%95%98%EA%B3%A0-%EA%B0%80%EB%B2%84%EB%A6%AC%EB%A9%B4-%EC%96%B4%EB%96%A1%ED%95%B4%EC%9A%94)

DuckOriDuck replied:
고쳐놓겠습니다

JIN-076 commented on src/main/java/io/dev/jobprep/common/base/BaseTimeEntity.java:
```diff
@@ -15,12 +15,12 @@ public class BaseTimeEntity {
     @CreatedDate
     @Column(name = "created_at")
-    private LocalDateTime createdAt;
+    protected LocalDateTime createdAt;
```
해당 클래스에 @Getter가 있음에도 해당 필드를 `protected` 로 변경하신 이유가 있을까요?
- `createdAt` 은 @CreatedDate 어노테이션으로 인해 데이터베이스 저장 시, 자동으로 저장됨
- `deletedAt` 은 `delete()` 를 활용해 값을 수정

그 이외에 해당 필드를 직접적으로 접근해서 값을 수정해야 할 일은 없기 때문에 이를 막기 위해 `private` 으로 작성해두었습니다.

DuckOriDuck replied:
후반에 수정하려고 했는데 중간에 놓쳤나봅니다. private으로 수정하겠습니다

JIN-076 commented on src/main/java/io/dev/jobprep/exception/code/ErrorCode401.java:
```diff
@@ -8,6 +8,8 @@ public enum ErrorCode401 implements ErrorCode {
     AUTH_MISSING_CREDENTIALS("E02-AUTH-001", "사용자의 인증 정보를 찾을 수 없습니다."),
     AUTH_TOKEN_EXPIRED("E02-AUTH-002", "토큰이 만료되었습니다."),
+
+    USER_ACCOUNT_DISABLED("E01-USER-001", "탈퇴한 유저입니다")
```
`E01` 은 `E02` 로 수정해주세요~

JIN-076 commented on src/main/java/io/dev/jobprep/users/application/UserService.java:
```diff
@@ -0,0 +1,74 @@
+    public DeleteUserAccountResponse deleteUserAcount(Long userId){
+        User userData = userRepository.findUserById(userId)
+                .orElseThrow(() -> new UserException(USER_NOT_FOUND));
+        return DeleteUserAccountResponse.from("회원탈퇴가 성공적으로 처리되었습니다.");
```
탈퇴 처리하는 로직은 일부러 빼신걸까요?

JIN-076 commented on src/main/java/io/dev/jobprep/users/domain/User.java:
```diff
@@ -0,0 +1,64 @@
+        this.penaltyUpdatedAt = penaltyUpdatedAt;
+    }
+
+    public void setDelete() {
```
해당 함수도 `delete()` 네이밍으로 변경해주세요~

DuckOriDuck committed:
[FIX] PR 수정사항
- 도메인 디렉토리 수정
- 메소드들 delete(), restore()로 명명
- BaseTimeEntity 에서 접근제한 private으로 수정
- 누락됐던 탈퇴처리 로직 재추가
- 마이그레이션 파일 V8로 변경

DuckOriDuck committed:
[FIX] 필드에서 time 지우고 메세지만 담도록 수정

DuckOriDuck committed:
[FIX] 에러코드 수정
